### PR TITLE
ci: Remove Python 3.8; add Python 3.12.

### PR DESCRIPTION
**Pull request by Niloth-p**
Fixes: Failing CI for Py 3.8
Fixes: #829
Fixes: #841 can be closed as well.

Removes Python 3.8.
Adds Py 3.11 Zulip 7.0 to matrix strategy.
Adds Python 3.12.

<details>
<summary>Self-review checklist</summary>

- [x] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [ ] Explains differences from previous plans (e.g., issue description).
- [x] Highlights technical choices and bugs encountered.
- [x] Calls out remaining decisions and concerns.
- [x] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [x] Each commit is a coherent idea.
- [x] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [ ] Visual appearance of the changes.
- [ ] Responsiveness and internationalization.
- [ ] Strings and tooltips.
- [ ] End-to-end functionality of buttons, interactions and flows.
- [ ] Corner cases, error conditions, and easily imagined bugs.
</details>

### CODE REVIEW

File: .github/workflows/zulip-ci.yml [hunk 1]
```diff
@@ -35,6 +29,18 @@ jobs:
             os: jammy
             legacy_client_interface: "6"
             server_version: refs/tags/6.0
+          # Debian 12 ships with Python 3.11.2.
+          - docker_image: zulip/ci:bookworm
+            name: Debian 12 (Python 3.11, backend + documentation)
+            os: bookworm
+            legacy_client_interface: "7"
+            server_version: refs/tags/7.0
```
Comment (Niloth-p) [inline .github/workflows/zulip-ci.yml:37]: Earliest major Zulip version that has the [same entry](https://github.com/zulip/zulip/blob/7.0/.github/workflows/zulip-ci.yml#L51-L54) in its matrix.

File: .github/workflows/zulip-ci.yml [hunk 2]
```diff
@@ -35,6 +29,18 @@ jobs:
...
+          # Ubuntu 24.04 ships with Python 3.12.3.
+          - docker_image: zulip/ci:noble
+            name: Ubuntu 24.04 (Python 3.12, backend)
+            os: noble
+            legacy_client_interface: "8"
+            server_version: refs/tags/8.5
```
Comment (Niloth-p) [inline .github/workflows/zulip-ci.yml:43]: Earliest minor Zulip version that [supports](https://github.com/zulip/zulip/blob/8.3/tools/lib/provision.py#L89) the platform is 8.3.

But, 8.3 and 8.4 are unable to use `pip` inside the venv, due to version conflicts related to `pkg_resources`. Only versions 8.5 onwards seem to be working.

<details>
<summary>Traceback</summary>

```
Traceback (most recent call last):
  File "/srv/zulip-py3-venv/bin/pip", line 5, in <module>
    from pip._internal.cli.main import main
  File "/srv/zulip-venv-cache/e2c2c053635e882ff393067adf908813bcad92ef/zulip-py3-venv/lib/python3.12/site-packages/pip/_internal/cli/main.py", line 10, in <module>
    from pip._internal.cli.autocompletion import autocomplete
  File "/srv/zulip-venv-cache/e2c2c053635e882ff393067adf908813bcad92ef/zulip-py3-venv/lib/python3.12/site-packages/pip/_internal/cli/autocompletion.py", line 9, in <module>
    from pip._internal.cli.main_parser import create_main_parser
  File "/srv/zulip-venv-cache/e2c2c053635e882ff393067adf908813bcad92ef/zulip-py3-venv/lib/python3.12/site-packages/pip/_internal/cli/main_parser.py", line 7, in <module>
    from pip._internal.cli import cmdoptions
  File "/srv/zulip-venv-cache/e2c2c053635e882ff393067adf908813bcad92ef/zulip-py3-venv/lib/python3.12/site-packages/pip/_internal/cli/cmdoptions.py", line 25, in <module>
    from pip._internal.cli.progress_bars import BAR_TYPES
  File "/srv/zulip-venv-cache/e2c2c053635e882ff393067adf908813bcad92ef/zulip-py3-venv/lib/python3.12/site-packages/pip/_internal/cli/progress_bars.py", line 12, in <module>
    from pip._internal.utils.logging import get_indentation
  File "/srv/zulip-venv-cache/e2c2c053635e882ff393067adf908813bcad92ef/zulip-py3-venv/lib/python3.12/site-packages/pip/_internal/utils/logging.py", line 18, in <module>
    from pip._internal.utils.misc import ensure_dir
  File "/srv/zulip-venv-cache/e2c2c053635e882ff393067adf908813bcad92ef/zulip-py3-venv/lib/python3.12/site-packages/pip/_internal/utils/misc.py", line 21, in <module>
    from pip._vendor import pkg_resources
  File "/srv/zulip-venv-cache/e2c2c053635e882ff393067adf908813bcad92ef/zulip-py3-venv/lib/python3.12/site-packages/pip/_vendor/pkg_resources/__init__.py", line 3254, in <module>
    @_call_aside
     ^^^^^^^^^^^
  File "/srv/zulip-venv-cache/e2c2c053635e882ff393067adf908813bcad92ef/zulip-py3-venv/lib/python3.12/site-packages/pip/_vendor/pkg_resources/__init__.py", line 3238, in _call_aside
    f(*args, **kwargs)
  File "/srv/zulip-venv-cache/e2c2c053635e882ff393067adf908813bcad92ef/zulip-py3-venv/lib/python3.12/site-packages/pip/_vendor/pkg_resources/__init__.py", line 3280, in _initialize_master_working_set
    tuple(
  File "/srv/zulip-venv-cache/e2c2c053635e882ff393067adf908813bcad92ef/zulip-py3-venv/lib/python3.12/site-packages/pip/_vendor/pkg_resources/__init__.py", line 3281, in <genexpr>
    dist.activate(replace=False)
  File "/srv/zulip-venv-cache/e2c2c053635e882ff393067adf908813bcad92ef/zulip-py3-venv/lib/python3.12/site-packages/pip/_vendor/pkg_resources/__init__.py", line 2788, in activate
    declare_namespace(pkg)
  File "/srv/zulip-venv-cache/e2c2c053635e882ff393067adf908813bcad92ef/zulip-py3-venv/lib/python3.12/site-packages/pip/_vendor/pkg_resources/__init__.py", line 2285, in declare_namespace
    _handle_ns(packageName, path_item)
  File "/srv/zulip-venv-cache/e2c2c053635e882ff393067adf908813bcad92ef/zulip-py3-venv/lib/python3.12/site-packages/pip/_vendor/pkg_resources/__init__.py", line 2202, in _handle_ns
    loader = importer.find_module(packageName)
             ^^^^^^^^^^^^^^^^^^^^
AttributeError: 'FileFinder' object has no attribute 'find_module'
```

</details>

File: .github/workflows/zulip-ci.yml [hunk 3]
```diff
@@ -35,6 +29,18 @@ jobs:
             os: jammy
             legacy_client_interface: "6"
             server_version: refs/tags/6.0
+          # Debian 12 ships with Python 3.11.2.
+          - docker_image: zulip/ci:bookworm
+            name: Debian 12 (Python 3.11, backend + documentation)
+            os: bookworm
+            legacy_client_interface: "7"
+            server_version: refs/tags/7.0
+          # Ubuntu 24.04 ships with Python 3.12.3.
```
Comment (Niloth-p) [inline .github/workflows/zulip-ci.yml:38]: Though, the Zulip workflow mentions [3.12.2](https://github.com/zulip/zulip/blob/8.3/.github/workflows/zulip-ci.yml#L57), the actual version used now seems to be 3.12.3.

File: zulip_bots/setup.py [hunk 4]
```diff
@@ -56,7 +56,7 @@
         "lxml",
         "BeautifulSoup4",
         "typing_extensions>=4.5.0",
-        'importlib-metadata >= 3.6; python_version  < "3.10"',
+        "importlib-metadata>=3.6",
```
Comment (Niloth-p) [inline zulip_bots/setup.py:59]: Fix for #829, included here as it is necessary for the Py 3.12 tests (but not for the 3.11 one).

**Review by Niloth-p — COMMENTED** [review]
Self-reviewed.

File: zulip_bots/setup.py (hunk 4 above)
Comment (timabbott) [reply zulip_bots/setup.py:59]: Hmm, isn't the point of #829 that we can just drop this dependency and use the built-in module in 3.10? Or do we need to wait until we drop 3.9 support to drop it?

**Comment by timabbott** [comment]
Seems OK, merged, thanks @Niloth-p!

File: zulip_bots/setup.py (hunk 4 above)
Comment (Niloth-p) [reply zulip_bots/setup.py:59]: Yes, we can't drop it before dropping 3.9 support.

**Pull request merged**